fs/zbd_zenfs.cc

```diff
@@ -653,42 +653,35 @@ unsigned int GetLifeTimeDiff(Env::WriteLifeTimeHint zone_lifetime, Env::WriteLif
   return LIFETIME_DIFF_NOT_GOOD;
 }
 
-Zone *ZonedBlockDevice::AllocateMetaZone(std::mutex& metadata_reset_mtx, std::condition_variable& metadata_cv) {
-
-  auto getResettedMetaZone = [this, &metadata_reset_mtx, &metadata_cv] () -> Zone* {
-    LatencyHistGuard guard(&meta_alloc_latency_reporter_);
-    meta_alloc_qps_reporter_.AddCount(1);
+Zone *ZonedBlockDevice::AllocateMetaZone() {
+  LatencyHistGuard guard(&meta_alloc_latency_reporter_);
+  meta_alloc_qps_reporter_.AddCount(1);
 
+  // Wait till there is an empty metazone
+  std::unique_lock<std::mutex> lk(metazone_reset_mtx_);
+  metazone_reset_cv_.wait(lk, [&] {
     for (const auto z : meta_zones) {
-      /* return the first zone that is not used and is empty */
-      if (!z->IsUsed()) {
-        if (!z->IsEmpty()) {
-          /* reset the zone asynchronously if it is not used and not empty.
-           * signal metadata condition variable once reset is finished*/
-          std::thread backgroundTask([z = std::move(z), &metadata_reset_mtx, &metadata_cv] {
-            std::unique_lock<std::mutex> lk(metadata_reset_mtx);
-            z->Reset();
-            metadata_cv.notify_one();
-          });
-          backgroundTask.detach();
-
-          /* continue the for-loop to find next non-used and empty metazone while async reset is happening */
+      if (!z->IsUsed() && z->IsEmpty()) {
+        return true;
+      }
+    }
+  });
+
+  // TODO(guokuankuan) we don't have to do the resetting here anymore
```

Comment:
why?

Reply:
@weilewei We depend on the background resetting, so the following lines don't need to reset again. But since we are refactoring the whole background metadata management, this should not be a problem anymore. We will erase all related codes in the future.